core/src/main/scala/kafka/log/Log.scala

```diff
@@ -357,12 +357,38 @@ class Log(@volatile var dir: File,
 
   def lastStableOffsetLag: Long = highWatermark - lastStableOffset
 
+  /**
+    * Fully materialize and return an offset snapshot including segment position info. This method will update
+    * the LogOffsetMetadata for the high watermark and last stable offset if they are message-only. Throws an
+    * offset out of range error if the segment info cannot be loaded.
+    */
   def offsetSnapshot: LogOffsetSnapshot = {
+
+    var highWatermark = _highWatermarkMetadata
+    if (highWatermark.messageOffsetOnly) {
+      lock.synchronized {
+        val fullOffset = convertToOffsetMetadataOrThrow(highWatermark.messageOffset)
+        _highWatermarkMetadata = fullOffset
+        highWatermark = _highWatermarkMetadata
+      }
+    }
+
+    val lastStable: LogOffsetMetadata = if (firstUnstableOffset.exists(_.messageOffsetOnly)) {
```

Comment:
I think we have the same "check then act" problem here. This is also a bit tricky because we are depending indirectly on the high watermark if the first unstable offset is not defined

```scala
var lso = lastStableOffsetMetadata
if (lso.messageOffsetOnly) {
  lock synchronized {
    firstUnstableOffset match {
      case None => highWatermark
      case Some(offsetMetadata) =>
        val fullOffset = convertToOffsetMetadataOrThrow(offsetMetadata.messageOffset)
        firstUnstableOffset = Some(fullOffset)
        lso = fullOffset
    }
  }
}
```